### PR TITLE
Feat/flavor v3 make disks recommended

### PR DESCRIPTION
**Pull request by garloff**
Notes on this PR:
* It consists of several steps (commits) that can be reviewed individually
  - Copy old v2 standard, add distinction b/w mandatory and recommended in text and in test tools (`--v3` aka `-3` option), add SSD flavors from 0110-v1 as mandatory.
* The last step could in theory also add the SSD flavors as recommended only
* In `--v3` mode, missing recommended flavors will create a warning, but no compliance error
* The test tools currently work with just one version of the required flavor file. This is currently broken by not dropping the SSD flavors from v2 and v1 tests -- this may need a different approach. (Maybe I would continue to use one version of the file with the required flavors, but annotate the new SSD ones, or create two versions or just hardcode dropping them without `--v3`. That latter is a bit hacky, admittedly.)
* wavestack and regio.digital both currently pass the v3 tests already (SSD flavors exist), ps fails by not having the new v2 names yet nor the SSD flavors.
* we would still have to decide when this standard becomes valid and when v1 and v2 are phased out.

### CODE REVIEW

**Comment by garloff** [comment]
This implements Ralfs suggestions from https://github.com/SovereignCloudStack/standards/issues/271#issuecomment-1538371315

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-name-check.py:632]: The `yamldict` is changed in place, so we could as well proceed as follows (modulo typos):

```
if prefer_old:
    for name_list in yamldict["SCS-Spec"].values():
        for i, name in enumerate(name_list):
            name_list[i] = new_to_old(name)
return yamldict["SCS-Spec"]["MandatoryFlavors"], yamldict["SCS-Spec"]["RecommendedFlavors"]
```

Then it's more obvious that the change is done in place, we get rid of the duplicated code, and the return statement is more telling (in my opinion).

File: Tests/iaas/flavor-naming/flavor-names-openstack.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-names-openstack.py:99]: This is `__dict__` check is quite hard to swallow for my taste. Could we instead use `not disk.parsed`? Would that do the trick as well?

Comment (mbuechse) [reply Tests/iaas/flavor-naming/flavor-names-openstack.py:99]:
```suggestion
            if not disk or not disk.parsed or disk.disktype != "s":
```

Comment (mbuechse) [reply Tests/iaas/flavor-naming/flavor-names-openstack.py:99]: On second thought, I think the list `scsV2Mand` should be constructed via list comprehension, and then the variable could be dropped for good. Something like so:

```python
def is_ssd_flavor(name):
    disk = fnmck.parsename(flv)[1]
    return disk and disk.parsed and disk.disktype == "s"

scsMandatory = [name for name in scsMandatory if not is_ssd_flavor(name)]
```

File: Tests/iaas/flavor-naming/flavor-names-openstack.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-names-openstack.py:214]: I would strongly advocate for trailing commas here and below!

File: Tests/scs-compatible.yaml
Comment (mbuechse) [inline Tests/scs-compatible.yaml:39]: I'm still puzzled by there being two yaml files. This PR only changes one of them. Why is that (I just don't know)?

**Review by mbuechse — COMMENTED** [review]
I have added a few remarks, nothing critical as far as I can tell.

File: Tests/iaas/flavor-naming/flavor-manager-input.py
Comment (fkr) [reply Tests/iaas/flavor-naming/flavor-manager-input.py:110]: I agree with @mbuechse and would suggest to remove it.

File: Tests/iaas/flavor-naming/flavor-manager-input.py
Comment (fkr) [reply Tests/iaas/flavor-naming/flavor-manager-input.py:113]: I agree with @mbuechse and would suggest to remove it.

Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-manager-input.py:112]: Done.

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-name-check.py]: Addressed in https://github.com/SovereignCloudStack/standards/pull/294/commits/5f0167b00dedc5b2df94c87784e0bbc863c71cd4.

File: Tests/iaas/flavor-naming/flavor-manager-input.py
Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-manager-input.py]: Addressed in https://github.com/SovereignCloudStack/standards/pull/294/commits/07bd711b0dc36a9841e2545f64a06c3783b6a330 with helper function.

File: Tests/iaas/flavor-naming/flavor-names-openstack.py
Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-names-openstack.py]: List comprehensions are definitely more readable, thanks.
Implemented in https://github.com/SovereignCloudStack/standards/pull/294/commits/642aa69165fc2712e0e807efc2e7d8f770454027

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-name-check.py]: Implemented in https://github.com/SovereignCloudStack/standards/pull/294/commits/02044a8b183a92de04df03d30dfee12ab7fa4cb7

File: Tests/iaas/flavor-naming/flavor-names-openstack.py
Comment (garloff) [reply Tests/iaas/flavor-naming/flavor-names-openstack.py]: No strong opinion. I tend to think it's a good habit to have these commas, so I just did it.
Implemented in https://github.com/SovereignCloudStack/standards/pull/294/commits/84fde967cbb3119ab87667a7bb5569b39b0d8269

**Comment by garloff** [comment]
I intend to redo the distinction between v3 flavors by changing the format of  `SCS-Spec.MandatoryFlavors.yaml`
to have sections `SCS-Spec.MandatoryFlavors`, `SCS-Spec.MandatoryFlavorsV3`, `SCS-Spec.RecommendedFlavors` and `SCS-Spec.RecommendedFlavorsV3` sections and handle the filtering in the loader, which get's a version number passed.
After that, I intend to remove the "work in progress" label and ask for more review.
Thanks for all the good input, @mbuechse.

File: Tests/scs-compatible.yaml
Comment (garloff) [reply Tests/scs-compatible.yaml:39]: Moved into a testing directory with https://github.com/SovereignCloudStack/standards/pull/294/commits/e5c0d811b07d76b2941da650267dc0f4944312d6.

**Comment by mbuechse** [comment]
> After that, I intend to remove the "work in progress" label and ask for more review.

Oh, I totally missed that label.

> I intend to redo the distinction between v3 flavors by changing the format of SCS-Spec.MandatoryFlavors.yaml

Now of course I'm wondering whether there is a more general (if not to say elegant) solution... You know, I did criticize some code duplication, but I'm not entirely against duplication if it serves a purpose. In that case, I can imagine using one file per version, even if that means some duplication, but then the code stays quite simple, and everything is very explicit.

**Comment by mbuechse** [comment]
Addendum: It would actually nicely mirror the situation with the standard texts (the markdown files in this repo), where we have one file per major version, regardless of duplication.

**Comment by garloff** [comment]
> > I intend to redo the distinction between v3 flavors by changing the format of SCS-Spec.MandatoryFlavors.yaml
> 
> Now of course I'm wondering whether there is a more general (if not to say elegant) solution... You know, I did criticize some code duplication, but I'm not entirely against duplication if it serves a purpose. In that case, I can imagine using one file per version, even if that means some duplication, but then the code stays quite simple, and everything is very explicit.

No, duplicating complex code is a bad idea. And the name parser is not trivial ...
One set of tools that can handle all versions (as long as we stick to a scheme that resembles what we currently have in v1/v2/v3).
We can discuss whether the `SCS-Spec.MandatoryFlavors.yaml` should be versioned. That would be harmless duplication.
I have decided against it for now (and instead have introduced `MandatoryFlavorsV3` list), but that's a discussion we can have and I'm open to be convinced differently. But no, the parser should be able to handle all versions IMVHO.

**Comment by garloff** [comment]
This is ready for review IMVHO.
I'd like to get this merged as Draft quickly.
So we have a solid proposal that is in line with current flavor name handling in SCS (v1, v2).

We may or may not stabilize it, depending on how the discussion goes w.r.t to the alternative proposal from @PSwatchmen at https://github.com/SovereignCloudStack/standards/issues/271#issuecomment-1536315678.

**Comment by garloff** [comment]
Sidenote: The MD link checker fails, as the linked file will only exist once the merge has been done.
So this failure can be ignored in judging the maturity of the PR, IMVHO.
Another sidenote: I have manually set DCO to pass -- the reason is that one commit from myself missed the needed signoff. As the commit history will be squashed and I know that I contributed this willingly and with all needed permissions, I can take care of it when the squashing happens. Thus no blocker either.

**Comment by mbuechse** [comment]
> No, duplicating complex code is a bad idea. And the name parser is not trivial ...

Just for the record (for posterity): We are in agreement here. My aim was to duplicate the yaml, not the Python file. Unfortunately, I didn't write sufficiently clearly.

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-name-check.py:665]:
```suggestion
        mand.extend(yamldict["SCS-Spec"].get("MandatoryFlavorsV3", ()))
```

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-name-check.py:670]:
```suggestion
        recd.extend(yamldict["SCS-Spec"].get("RecommendedFlavorsV3", ()))
```

File: Tests/iaas/flavor-naming/flavor-name-check.py
Comment (mbuechse) [inline Tests/iaas/flavor-naming/flavor-name-check.py:731]: Just a bit more "pythonic":
```suggestion
    if argv and argv[0] == "-i":
```

**Review by tibeer — APPROVED** [review]
LGTM

**Comment by tibeer** [comment]
@garloff openstack flavor manager has been altered: https://github.com/osism/openstack-flavor-manager/pull/11

**Comment by garloff** [comment]
Great, thanks Tim

**Pull request merged**
